### PR TITLE
IGNITE-3464

### PR DESCRIPTION
**Pull request by BiryukovVA**
Fixed bug - race between partition exchange and prepare/finish requests.

### CODE REVIEW

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/TxOptimisticOnPartitionExchange.java [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+            awaitPartitionMapExchange();
+
+            msgInterception = false;
+
+            IgniteCache<Object, Object> cacheStartedNode = ignite.cache(CACHE_NAME);
+
+            try (Transaction tx = ignite.transactions().txStart(PESSIMISTIC, REPEATABLE_READ)) {
+                Set<Object> keys = cacheStartedNode.getAll(txValues.keySet()).keySet();
+
+                assertEquals(txValues.keySet(), new TreeSet<>(keys));
```
[the remark this answers is not in the document]
Comment (BiryukovVA) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/TxOptimisticOnPartitionExchange.java]: getAll() methot doesn't return null values.

[the remark this answers is not in the document]
Comment (SomeFire) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/TxOptimisticOnPartitionExchange.java]: Ok, my bad.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/TxOptimisticOnPartitionExchange.java [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+    }
+
+    /**
+     * @param isolation {@link TransactionIsolation}.
+     * @param txInitiatorPrimary False If the transaction does not use the keys of the node that initiated it.
+     * @throws Exception If failed.
+     */
+    @SuppressWarnings("unchecked")
+    public void doTest(final TransactionIsolation isolation, boolean txInitiatorPrimary) throws Exception {
+        final CountDownLatch txStarted = new CountDownLatch(1);
```
[the remark this answers is not in the document]
Comment (BiryukovVA) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/TxOptimisticOnPartitionExchange.java]: Actually, it doesn't matter when to call countDown method right after tx.start or after putAll.
I can move it right after tx start.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/TxOptimisticOnPartitionExchange.java [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+        }
+
+        TestCommunicationSpi.init();
+
+        msgInterception = true;
+
+        IgniteInternalFuture<Object> fut = GridTestUtils.runAsync(new Callable<Object>() {
+            @Override public Object call() {
+                try (Transaction tx = ignite(0).transactions().txStart(OPTIMISTIC, isolation)) {
+                    log().info(">>> TX started.");
```
[the remark this answers is not in the document]
Comment (BiryukovVA) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/TxOptimisticOnPartitionExchange.java]: done.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/TxOptimisticOnPartitionExchange.java [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+import static org.apache.ignite.transactions.TransactionConcurrency.OPTIMISTIC;
+import static org.apache.ignite.transactions.TransactionConcurrency.PESSIMISTIC;
+import static org.apache.ignite.transactions.TransactionIsolation.READ_COMMITTED;
+import static org.apache.ignite.transactions.TransactionIsolation.REPEATABLE_READ;
+import static org.apache.ignite.transactions.TransactionIsolation.SERIALIZABLE;
+
+/**
+ *
+ */
+public class TxOptimisticOnPartitionExchange extends GridCommonAbstractTest {
```
[the remark this answers is not in the document]
Comment (BiryukovVA) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/TxOptimisticOnPartitionExchange.java]: done.

**Pull request closed without merging**